src/com/can/CANLayer.cpp

```diff
@@ -0,0 +1,157 @@
+#include <linux/can.h>
+#include <string>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+
+#include "CANLayer.h"
+#include "basecommfb.h"
+#include "comtypes.h"
+#include "datatype.h"
+#include "devlog.h"
+#include "forte_any.h"
+#include "forte_byte.h"
+
+using namespace forte::com_infra;
+
+CCANComLayer::CCANComLayer(CComLayer *paUpperLayer, CBaseCommFB *paComFB) :
+	CComLayer(paUpperLayer, paComFB) {}
+
+EComResponse CCANComLayer::openConnection(char *paLayerParameter) {
+	EComResponse eRetVal = e_InitInvalidId;
+	CParameterParser parser(paLayerParameter, ';', eCANComParamterAmount);
+	if (eCANComParamterAmount != parser.parseParameters()) {
+		return eRetVal;
+	}
+
+	setupCANInternals(parser);
+
+	switch (mFb->getComServiceType()) {
+	case e_Server:
+		DEVLOG_DEBUG("[CAN Layer] the layer only supports the PUB/SUB "
+					 "pattern. Please use a Publisher block to send data.");
+		eRetVal = e_InitTerminated;
+		break;
+	case e_Client:
+		DEVLOG_DEBUG("[CAN Layer] the layer only supports the PUB/SUB "
+					 "pattern. Please use a Subscribe block to receive data.");
+		eRetVal = e_InitTerminated;
+		break;
+	case e_Publisher:
+		// is handled via sendData
+		break;
+	case e_Subscriber:
+		// is handled via recvData
+		break;
+	}
+
+	DEVLOG_DEBUG("[CAN Layer] Opening connection\n");
+
+	CFDSelectHandler::TFileDescriptor fd = socket(PF_CAN, SOCK_RAW, CAN_RAW);
+
+	if (CFDSelectHandler::scmInvalidFileDescriptor == fd) {
+		DEVLOG_DEBUG("[CAN Layer] Socket can't be created.\n");
+		return forte::com_infra::e_InitInvalidId;
+	}
+
+	mCANSocket = fd;
+	std::strcpy(mIfr.ifr_name, mCANParam.interfaceName.c_str());
+	ioctl(mCANSocket, SIOCGIFINDEX, &mIfr);
+	mAddr.can_family = AF_CAN;
+	mAddr.can_ifindex = mIfr.ifr_ifindex;
+
+	if (bind(mCANSocket, (struct sockaddr *)&mAddr, sizeof(mAddr)) < 0) {
+		DEVLOG_DEBUG("[CAN Layer] Binding of the socket unsuccessful.\n");
+		return eRetVal;
+	}
+
+	this->getExtEvHandler<CFDSelectHandler>().addComCallback(mCANSocket, this);
+	DEVLOG_DEBUG("[CAN Layer] Callback handler added.\n");
+	eRetVal = e_InitOk;
+
+	return eRetVal;
+}
+
+CCANComLayer::~CCANComLayer() { closeConnection(); }
+
+void CCANComLayer::closeConnection() {
+	if (mCANSocket != CFDSelectHandler::scmInvalidFileDescriptor) {
+		DEVLOG_DEBUG("[CAN Layer] Closing socket\n");
+		close(mCANSocket);
+	}
+	DEVLOG_DEBUG("[CAN Layer] Unregister layer\n");
+	this->getExtEvHandler<CFDSelectHandler>().removeComCallback(mCANSocket);
+}
+
+EComResponse CCANComLayer::sendData(void *paData, unsigned int paSize) {
+	auto *raw_data = static_cast<uint8_t *>(
+		(&static_cast<CIEC_ANY **>(paData)[0]->unwrap())->getDataPtr());
+	size_t data_size = paSize;
+
+	if (data_size > 8) {
+		DEVLOG_WARNING("[CAN Layer] The max. size of a CAN frame is: %u. Data "
+					   "will be lost.",
+					   CAN_MAX_DLEN);
+	}
+
+	DEVLOG_DEBUG("[CAN Layer] Size of data is: %u\n", data_size);
+
+	size_t offset = 0;
+	while (offset < data_size) {
+		mFrame.can_id = mCANParam.CANId;
```

Comment:
the CANId was set when opening the connection, so you could set it directly into mFrame.can_id there and get rid of the `mCANParam.CANId`